src/client/ua_client_connect.c

```diff
@@ -1709,6 +1709,72 @@ UA_Client_connectSecureChannel(UA_Client *client, const char *endpointUrl) {
     return client->connectStatus;
 }
 
+static void
+activateSessionSync(UA_Client *client) {
+    UA_LOCK_ASSERT(&client->clientMutex, 1);
+    
+    UA_DateTime now = UA_DateTime_nowMonotonic();
+    UA_DateTime maxDate = now + ((UA_DateTime)client->config.timeout * UA_DATETIME_MSEC);
+
+    // reset state to created
+    client->sessionState = UA_SESSIONSTATE_CREATED;
```

Comment:
Check whether the session is either already created or already activated.
Because you could call this method when there is no session at all.
(Connection with only a SecureChannel.)

You could do the test within `activateSessionAsync`.
With logging and a return statuscode.
This statuscode (if bad) should get returned from UA_Client_activateSession and UA_Client_activateSessionAsync.

Note that it does not impact the connectStatus.
As the SecureChannel might still be good.